### ISSUE TITLE
apply() modifies the patch when working with arrays

### ISSUE THREAD

**Issue by fhuonder** [pikaju/dart-json-patch#1]
Hi,

I am encountering a bug when working with arrays.

Assume the following patch (in object form)
```json
[
  {"op": "add", "path": "/a/b/c", "value": []},
  {"op": "add", "path": "/a/b/c/-", "value": {"something":true}}
]
```
The procedure is:
```dart
var patch = ...; //the patch from above
JsonPatch.apply(myDocument, patch);
jsonEncode(patch);
```

In the encoded JSON on line 3 the patch has become
```json
[
  {"op": "add", "path": "/a/b/c", "value": [{"something":true}]},
  {"op": "add", "path": "/a/b/c/-", "value": {"something":true}}
]
```

I assume that the source of the problem is in the `_addChild` method.
There, for the first patch-entry the value of the patch is copied into the target object. In this case, the value is an empty list.
The second patch entry appends to the same list and in the `_addChild` method for that case the value is appended to the list in the target object. But because the list is not copied only its reference is put into the target object. This leads to the problem that the patch and target object contain the same list instance.

As I understand it the value in the patch can be a List, Map or primitive value. In case it is a Map or List, those objects must be copied (deep copy needed) instead of simply assigned.

Can you confirm? Could you provide a patch?

Regards,
Florian

**Comment by pikaju** [comment]
Hey, that is indeed an interesting bug you found. I haven't considered that. Probably an easy fix though, I'll look into it.

> Could you provide a patch?

I see what you did there.

**Comment by pikaju** [comment]
Okay, should be fixed in version 2.0.1, could you check if it works?

**Comment by fhuonder** [comment]
Seems to work.
Thanks
